---
title: "Progeny"
author: "Christian Heyer"
date: "8/23/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, 
                      fig.width = 12, fig.height = 12)
library(progeny)
library(DESeq2)
library(BiocParallel)
library(readr)
library(magrittr)
library(tidyverse)
library(RColorBrewer)
library(ggplot2)
library(decoupleR)
library(ComplexHeatmap)
if (!require(RNAscripts)) {
  devtools::install("./scripts/RNAscripts", upgrade = "never")
}
library(RNAscripts)
if (exists("snakemake")) {
  dds_path <- snakemake@input[['dds_obj']]
  diffexp_tb_path <- snakemake@input[['table']]
  fpkm_path <- snakemake@input[['fpkm']]
  cond_id <- snakemake@wildcards[["condition"]]
  contrast_groups <- snakemake@wildcards[["contrast"]]
  s_groups <- snakemake@params[["s_groups"]]
  register(MulticoreParam(snakemake@threads))
  contrast_name <- contrast_groups
  plot_path <- snakemake@params[["plot_path"]]
  comp_groups <- snakemake@config[["comp_groups"]]
  color_scheme <- snakemake@config[["group_colors"]][[cond_id]]
  organism <- snakemake@config[["organism"]]
} else {
  the_yaml <- yaml::read_yaml("../../configs/VascAge_Apelin_config.yaml")
  BASE_ANALYSIS_DIR <- ""
  cond_id <- "condition"
  comp_id <- "AECplusA-vs-AEC"
  dds_path <- file.path(paste0(BASE_ANALYSIS_DIR), "deseq2/all.rds")
  diffexp_tb_path <- file.path(BASE_ANALYSIS_DIR, 
                               glue::glue("results/diffexp/{comp_id}.diffexp.tsv"))
  fpkm_path <- file.path(BASE_ANALYSIS_DIR, "fpkm/all.tsv")
  contrast_groups <- c("AECplusA", "AEC")
  plot_path <- "."
  register(SerialParam())
  s_groups<- c("AECplusA", "AEC")
  contrast_name <- glue::glue("{contrast_groups[[1]]} vs {contrast_groups[[2]]}")
  comp_groups <- the_yaml$comp_groups
  color_scheme <- the_yaml$group_colors$condition
  organism <- "Mus musculus"
  
}
dir.create(plot_path, recursive = T)

```

## Progeny Analysis

Progeny analyses gene signatures and can score pathway activity of signalling pathways 
for a set of samples and also analyze the results from differential expression experiments.

```{r read data}

dds_obj <- readRDS(dds_path)
diffexp_tb <- read_tsv(diffexp_tb_path,  col_names = c('gene_id', 
                                                 'baseMean',
                                                 'logFoldChange',
                                                 'lfcSE', 'stat',
                                                 'pvalue', 'padj'), 
                                   skip = 1)
# Issue: we cant use vst for a small test dataset but getVarianceStabilzedData Works
Normalized_counts <- getVarianceStabilizedData(dds_obj)
#Normalized_counts<- assay(vst(dds_obj, blind = FALSE))
fpkm <- read_tsv(fpkm_path)
filer <- fpkm %>% dplyr::filter(gene %in% rownames(Normalized_counts)) %>%  dplyr::filter(!duplicated(gname))

joined_df <- join_tables(diffexp_tb,filer) %>% dplyr::filter(!duplicated(gname))
Normalized_counts <- Normalized_counts[filer$gene,]

rownames(Normalized_counts) <- filer$gname

joined_df %>% dplyr::select(gname, stat) %>% dplyr::filter(!is.na(stat)) %>%
  column_to_rownames(var = "gname") %>% as.matrix() -> diffexp_matrix
write("finished parsing", file =stderr())
```

## Progeny results over all samples

The following plot shows the heatmap of Progeny score for gene expression normalized
via DESeq2 for each sample. 

```{r pressure, echo=FALSE, fig.width=7}
org_name <- get_organism_omnipath_name(organism) 

prog_net <- decoupleR::get_progeny(organism = org_name, top = 100)


PathwayActivity_counts <- progeny(Normalized_counts, scale = T, organism = "Mouse", top = 1000, verbose = F, perm = 1) 
# Test case has some nans in progeny, these are set to 0
PathwayActivity_counts[is.nan(PathwayActivity_counts)] <- 0

Activity_counts <- as.vector(PathwayActivity_counts)


ann_col <- colData(dds_obj) %>% as_tibble()%>% pull("condition") %>% fct_relevel(names(color_scheme))

top_anno <- ComplexHeatmap::HeatmapAnnotation(sample = ann_col, 
                                                col = list("sample" = as_vector(color_scheme)))
progeny_hmap <- ComplexHeatmap::Heatmap(t(PathwayActivity_counts), 
                                          top_annotation = top_anno,
                                          clustering_distance_columns = "euclidean",
                                          clustering_method_columns = "average",
                                        show_column_names = F, name = )
save_cheatmap_svg(x = progeny_hmap, filename = file.path(plot_path, "progeny_hmap.svg"))
progeny_hmap
```

```{r}
library(broom)
as_tibble(PathwayActivity_counts) %>% t() -> transposed_pact

kruskal_res<- apply(transposed_pact, 1, kruskal.test, g = as.factor(ann_col))

kruskal_test <- purrr::map_df(kruskal_res, tidy,.id = "pathway") %>% mutate(padj = p.adjust(p.value, method = "BH")) %>% arrange(desc(statistic))
```

## Per Pathway activity plots

Here we first visualize the distribution of pathway activities between the 
comparison given. A kruskal wallis test is run to test if samples originate
from one or from multiple distributions. 


```{r}
plot_activity <- as_tibble(PathwayActivity_counts) %>% add_column(condition = fct_relevel(ann_col, s_groups))

for(pathway in kruskal_test$pathway) {
  pval <- kruskal_test %>% dplyr::filter(pathway == !!pathway) %>% pull(padj) 
  p <- ggplot(plot_activity, aes(x = condition, y = !!sym(pathway), fill = condition)) + 
    geom_boxplot(outlier.shape = NA, color = "black") + geom_jitter(width = 0.1) + theme_bw() + 
    scale_fill_manual(values = as_vector(color_scheme)) +
    geom_hline(yintercept = 0) + ggtitle(pathway) + labs(caption= paste0("kruskal adj.p = ", round(pval, 4))) + ylab("progeny score") + 
    ggsignif::geom_signif(comparisons = comp_groups, 
                          step_increase = 0.06, map_signif_level = T,show.legend = , color = "black")
  
  ggsave(filename = file.path(plot_path, glue::glue("{pathway}_boxplot.svg")))
  plot(p)
}
```

### Enrichment based off of differential gene expression analysis

Here we calculate the pathway enrichment based on the test statistic from DESeq2
to infer pathway changes between the comparison groups. 

```{r}
PathwayActivity_zscore <- progeny(diffexp_matrix, 
    scale=TRUE, organism="Mouse", top = 100, perm = 10000, z_scores = TRUE, verbose = T) %>%
    t()
colnames(PathwayActivity_zscore) <- "NES"
```

```{r}
PathwayActivity_zscore_df <- as.data.frame(PathwayActivity_zscore) %>% 
    rownames_to_column(var = "Pathway") %>%
    dplyr::arrange(NES) %>%
    dplyr::mutate(Pathway = factor(Pathway))

NES_plot <- ggplot(PathwayActivity_zscore_df,aes(x = reorder(Pathway, NES), y = NES)) + 
    geom_bar(aes(fill = NES), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 45, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) +
    xlab("Pathways") + ggtitle(contrast_name)

ggsave(filename = file.path(plot_path, "NES_plot.svg"), NES_plot)
NES_plot
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r, eval= ifelse(nrow(Normalized_counts) > 1000, TRUE, FALSE)} 

prog_matrix <- getModel(org_name, top=100) %>% 
    as.data.frame()  %>%
    tibble::rownames_to_column("GeneID")

ttop_KOvsWT_df <- diffexp_matrix %>% 
    as.data.frame() %>% 
    tibble::rownames_to_column("GeneID")


scatter_plots <- progeny::progenyScatter(df = ttop_KOvsWT_df, 
    weight_matrix = prog_matrix, 
    statName = "stat", verbose = T)

yeet <- purrr::map2(scatter_plots[[1]], names(scatter_plots[[1]]), function(x, n) ggsave(filename = file.path(plot_path, paste0(n,"_scatter.svg")), plot = x))
purrr::map(scatter_plots[[1]], plot)
```

## DecoupleR

With the decoupleR prviding a new more flexible interface for 
Enrichment analyses. To retain comparisons with the original progeny method we 
reproduce most of the plots above but now use the newer decoupleR::run_wmean for
inference in the progeny network. 

```{r, echo = F}
net <- get_progeny(organism = org_name, top = 100)

minsize = ifelse(nrow(Normalized_counts) > 1000, 5, 2)
times <- ifelse(nrow(Normalized_counts) > 1000, 1000, 100)

sample_acts <- decoupleR::run_wmean(mat=Normalized_counts, net=net, .source='source', .target='target',
                  .mor='weight', times = times, minsize = minsize)
```

```{r}
# Transform to wide matrix
sample_acts_mat <- sample_acts %>%
  filter(statistic == 'norm_wmean') %>%
  pivot_wider(id_cols = 'condition', names_from = 'source',
              values_from = 'score') %>%
  column_to_rownames('condition') %>%
  as.matrix()

# Scale per sample
sample_acts_mat <- scale(sample_acts_mat) %>% t()
sample_acts_mat <- sample_acts_mat[,rownames(dds_obj@colData)]


ha <- HeatmapAnnotation(group = dds_obj@colData$condition,
                        col = list(group = 
                          as_vector(color_scheme))
)

# Plot

ComplexHeatmap::Heatmap(sample_acts_mat, 
  top_annotation =  ha, clustering_method_columns = "average",
  clustering_method_rows = "average", show_column_names = F) 



```

### Single sample mean plots using decoupler

```{r, eval = ifelse(nrow(Normalized_counts) > 1000, TRUE, FALSE)}
as_tibble(sample_acts_mat)  -> out_yeet

yeet<- apply(sample_acts_mat, 1, kruskal.test, g = as.factor(ann_col))

kruskal_test <- purrr::map_df(yeet, tidy,.id = "pathway") %>% mutate(padj = p.adjust(p.value, method = "BH")) %>% arrange(desc(statistic))
plot_activity <- sample_acts_mat %>% t()  %>% as.data.frame()
plot_activity[,cond_id] <- ann_col

for(pathway in kruskal_test$pathway) {
  pval <- kruskal_test %>% dplyr::filter(pathway == !!pathway) %>% pull(padj) 
  p <- ggplot(plot_activity, aes(x = condition, y = !!sym(pathway), fill = condition)) + 
    geom_boxplot(outlier.shape = NA, color = "black") + geom_jitter(width = 0.1) + theme_bw() + 
    scale_fill_manual(values = as_vector(color_scheme)) +
    geom_hline(yintercept = 0) + ggtitle(pathway) + labs(caption= paste0("kruskal adj.p = ", round(pval, 4))) + ylab("progeny score") + 
    ggsignif::geom_signif(comparisons = comp_groups, 
                          step_increase = 0.06, map_signif_level = T,show.legend = , color = "black")
  
  ggsave(filename = file.path(plot_path, glue::glue("{pathway}_decoupler_boxplot.svg")))
  plot(p)
}

```

```{r}
deg <- joined_df %>% dplyr::select(gname, stat) %>% dplyr::filter(!is.na(stat)) %>%
  column_to_rownames(var = "gname") %>% as.matrix() -> diffexp_matrix
design <- dds_obj@colData
contrast_acts <- run_wmean(mat=deg, net=net, .source='source', .target='target',
                  .mor='weight', times = times, minsize = minsize)
# Filter norm_wmean
f_contrast_acts <- contrast_acts %>%
  filter(statistic == 'norm_wmean')

# Plot
ggplot(f_contrast_acts, aes(x = reorder(source, score), y = score)) + 
    geom_bar(aes(fill = score), stat = "identity") +
    scale_fill_gradient2(low = "darkblue", high = "indianred", 
        mid = "whitesmoke", midpoint = 0) + 
    theme_minimal() +
    theme(axis.title = element_text(face = "bold", size = 12),
        axis.text.x = 
            element_text(angle = 45, hjust = 1, size =10, face= "bold"),
        axis.text.y = element_text(size =10, face= "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank()) + 
    xlab("Pathways")
```

```{r}

plot_pathway <- function(net, pathway, deg) {
  df <- net %>%
    filter(source == pathway) %>%
    arrange(target) %>%
    mutate(ID = target, color = "3") %>%
    column_to_rownames('target')
  inter <- sort(intersect(rownames(deg),rownames(df)))
  df <- df[inter, ]
  df['t_value'] <- deg[inter, ]
  df <- df %>%
    mutate(color = if_else(weight > 0 & t_value > 0, '1', color)) %>%
    mutate(color = if_else(weight > 0 & t_value < 0, '2', color)) %>%
    mutate(color = if_else(weight < 0 & t_value > 0, '2', color)) %>%
    mutate(color = if_else(weight < 0 & t_value < 0, '1', color))
  
  ggplot(df, aes(x = weight, y = t_value, color = color)) + geom_point() +
    scale_colour_manual(values = c("red","royalblue3","grey")) +
    ggrepel::geom_label_repel(aes(label = ID)) + 
    theme_minimal() +
    theme(legend.position = "none") +
    geom_vline(xintercept = 0, linetype = 'dotted') +
    geom_hline(yintercept = 0, linetype = 'dotted') +
    ggtitle(pathway)
}
purrr::map(unique(net$source), plot_pathway,net = net,deg = deg)
```