---
title: "SALMONTE analysis"
author: "Christian Heyer"
date: "5/19/2021"
output: powerpoint_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = F, message = F)
library(DESeq2)
library(tidyverse)
library(tximport)
```


## Overview

- Methods and Approaches
- SALMONTE
- Preliminary results
- Limits and what now


## Methods and Approaches

- Two distinction in analysis types:
  - Family and Locus based
- Problem: Assigning short reads to 
- Most methods are based of EM-algorithms
  - improves read assignment to TE regions

## Tool choice

- REdiscoverTE (not really available)
- telescope (no mouse annotations)
- SALMONTE supports Mus musculus
- SQuire: requires realignment of entire dataset with star

## SalmonTE

SalmonTE (https://www.tandfonline.com/doi/full/10.1080/15592294.2019.1634975) is 
a salmon based pseudoaligner for analysis of transposable elements in the genome. 
An expression element is calculated for each transposable element given
via a predefined set of elements. 



```{r cars}
base_path <- "/icgc/dkfzlsdf/analysis/VascularAging/rna_sequencing/Vasc_age2020/SALMON_TE/"
expr_csv <- file.path(base_path, "all_samples/EXPR.csv")
clades_csv <- file.path(base_path, "all_samples/clades.csv")
mapping_info <- file.path(base_path, "all_samples/MAPPING_INFO.csv")

units_tsv <- "../../data/Vasc_age2020/units_vascage.tsv"
samples_tsv <- "../../data/Vasc_age2020/Vascage_samples.tsv"
```


```{r read and format}

units_info <- readr::read_tsv(units_tsv)
samples_info <- readr::read_tsv(samples_tsv)

meta_data <- inner_join(units_info, samples_info) %>% mutate(col_name = str_remove(fq1, ".fastq.gz"))

files <- file.path(base_path, "all_samples", meta_data$col_name, "quant.sf")
names(files) <- glue::glue("{meta_data$sample}_{meta_data$unit}")

count_m <- tximport(files,type = "salmon", txOut = TRUE)

col_data <- DataFrame(meta_data)
rownames(col_data) <- col_data$col_name

ddsTxi <- DESeqDataSetFromTximport(count_m, 
                                   colData = col_data,
                                   design = ~ condition,
                                  )

clades <- readr::read_csv(clades_csv)
```
## Including Plots

You can also embed plots, for example:

```{r rundeseq, echo=FALSE}

contrast_groups <- list("18m_vs_d0" = c("18m-lung", "d0-lung"),
                        "d15_vs_d0" = c("d15-lung", "d0-lung"),
                        "d22_vs_d0" = c("d22-lung", "d0-lung"),
                        "d36_vs_d0" = c("d36-lung", "d0-lung"),
                        "d36_vs_d15" = c("d36-lung", "d15-lung"))

dds <- DESeq(ddsTxi)

shrink_and_res <- function(dds, c_group) {
  results(dds, contrast = c("condition", c_group)) 
}
#resLFC <- lfcShrink(dds, coef="condition_d0.lung_vs_18m.lung")
#res <- results(dds, contrast = c("condition", "18m-lung", "d0-lung"))

all_res <- purrr::map(contrast_groups, shrink_and_res, dds = dds)

filter_sig <- function(res_table) {
   as_tibble(res_table,rownames = "gene_names")  %>% filter(pvalue < 0.05 & abs(log2FoldChange) > 0.3)
}

sig_res <- purrr::map(all_res, filter_sig)

```


```{r}
openxlsx::write.xlsx(sig_res, file.path(base_path, "sig_TE_transkripts.xlsx"))
```

## Heatmap

```{r, fig.height = 8, fig.width = 10, fig.cap = "Heatmap of scaled reptitive element expression"}
library(ComplexHeatmap)
rlog_dds <- rlog(dds)

expr_values <- assay(rlog_dds) 
row_clustering <- fastcluster::hclust(dist(expr_values))
clade_index <- !is.nan(rowSums(expr_values)) 
expr_values <- expr_values[clade_index,]
norm_expr_values <- expr_values %>% t() %>% scale(center = T, scale = T) %>% t()

c_ha <- HeatmapAnnotation(group = col_data$condition,
                          rep = col_data$unit, col = list(group = setNames(RColorBrewer::brewer.pal(length(unique(col_data$condition)),"Accent"),
                                                                            nm = unique(col_data$condition))))
r_ha <- rowAnnotation(clade_class = clades[clade_index,]$clade)
ComplexHeatmap::Heatmap(norm_expr_values, top_annotation = c_ha, 
                        right_annotation = r_ha, 
                        show_row_names = F, col = viridis::viridis(100),
                        clustering_method_columns = "ward.D", show_column_names = F, 
                        cluster_rows = row_clustering
                        )
                        

```

````{r, fig.height = 8, fig.width = 10, fig.caption = "Volcano plot of differentially expressed genes"}
library(EnhancedVolcano)

do_enhanced_vol <- function(res, title) {
    EnhancedVolcano(res,
    lab = rownames(res),
    x = 'log2FoldChange',
    y = 'pvalue', title = title,
    pCutoff = 0.05
    )
}

plots <- purrr::map2(all_res, names(all_res), do_enhanced_vol)

plots
```
